bundler/lib/bundler/definition.rb

```diff
@@ -320,38 +320,26 @@ def groups
       dependencies.map(&:groups).flatten.uniq
     end
 
-    def lock(file, preserve_unknown_sections = false)
-      return if Definition.no_lock
-
-      contents = to_lock
-
-      # Convert to \r\n if the existing lock has them
-      # i.e., Windows with `git config core.autocrlf=true`
-      contents.gsub!(/\n/, "\r\n") if @lockfile_contents.match?("\r\n")
-
-      if @locked_bundler_version
-        locked_major = @locked_bundler_version.segments.first
-        current_major = bundler_version_to_lock.segments.first
-
-        updating_major = locked_major < current_major
-      end
+    def lock(file_or_preserve_unknown_sections = false, preserve_unknown_sections_or_unused = false)
+      if [true, false, nil].include?(file_or_preserve_unknown_sections)
+        target_lockfile = lockfile || Bundler.default_lockfile
```

Comment:
Regression: this causes bundler/inline, which had a `nil` lockfile previously, to fall back to the default lockfile—which may exist—and overwrite it with the inline definition's lock.

This is a common pattern with bin/some-executable that executes with an inline bundle distinct from the main bundle.

To reproduce:
```bash
mkdir inline-lockfile-bug
cd inline-lockfile-bug
touch Gemfile.lock
chmod u-w Gemfile.lock  # to demonstrate that the inline bundle tries to overwrite
ruby -v; gem -v; bundle -v
ruby 3.3.5 (2024-09-03 revision ef084cc8f4) [arm64-darwin24]
3.3.22
Bundler version 2.5.19
ruby -rbundler/inline -e 'gemfile {}'
/Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler/definition.rb:547:in `initialize': Permission denied @ rb_sysopen - /Users/jeremy/inline-lockfile-bug/Gemfile.lock (Errno::EACCES)
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler/definition.rb:547:in `open'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler/definition.rb:547:in `block in write_lock'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler/shared_helpers.rb:105:in `filesystem_access'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler/definition.rb:546:in `write_lock'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler/definition.rb:337:in `lock'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler/runtime.rb:98:in `lock'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler/runtime.rb:35:in `setup'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler/inline.rb:64:in `block (2 levels) in gemfile'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler/settings.rb:158:in `temporary'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler/inline.rb:50:in `block in gemfile'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler.rb:429:in `block in with_unbundled_env'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler.rb:684:in `with_env'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler.rb:429:in `with_unbundled_env'
	from /Users/jeremy/.local/share/mise/installs/ruby/3.3.5/lib/ruby/gems/3.3.0/gems/bundler-2.5.19/lib/bundler/inline.rb:42:in `gemfile'
	from -e:1:in `<main>'
```

Reply:
Hei @jeremy!

You're totally right about this regression. It did not manifest until Bundler 2.5.19 because `bundler/inline` used to make `Definition#lock` a noop until https://github.com/rubygems/rubygems/commit/0b7be7bb770538fc26238dbaf055be9b06062ef8.

We got this reported last week and will be fixing it through https://github.com/rubygems/rubygems/pull/8055. I hope to do a quick release to patch this today or tomorrow.

Reply:
I released Bundler 2.5.20 with this fix!

Reply:
Nice!